File: week_2/code/day_9/b_lca_bst.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        curr = root
        
        while curr:
            # If both p and q are greater than the current node's value, LCA is in the right subtree
            if p.val > curr.val and q.val > curr.val:
                curr = curr.right
            # If both p and q are less than the current node's value, LCA is in the left subtree
            elif p.val < curr.val and q.val < curr.val:
                curr = curr.left
            # We found the split point (or one of the nodes is the LCA of the other)
            else:
                return curr

# Example Usage
if __name__ == "__main__":
    sol = Solution()
    
    #      6
    #    /   \
    #   2     8
    #  / \   / \
    # 0   4 7   9
    #    / \
    #   3   5
    
    n0 = TreeNode(0)
    n2 = TreeNode(2)
    n3 = TreeNode(3)
    n4 = TreeNode(4)
    n5 = TreeNode(5)
    n6 = TreeNode(6)
    n7 = TreeNode(7)
    n8 = TreeNode(8)
    n9 = TreeNode(9)
    
    n6.left, n6.right = n2, n8
    n2.left, n2.right = n0, n4
    n4.left, n4.right = n3, n5
    n8.left, n8.right = n7, n9
    
    # LCA of 2 and 8 is 6
    print(sol.lowestCommonAncestor(n6, n2, n8).val)
    
    # LCA of 2 investments 4 is 2
    print(sol.lowestCommonAncestor(n6, n2, n4).val)

File: week_2/code/day_9/b_lca_bst.js
// Definition for a binary tree node.
function TreeNode(val) {
    this.val = val;
    this.left = this.right = null;
}

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function (root, p, q) {
    let curr = root;

    while (curr) {
        if (p.val > curr.val && q.val > curr.val) {
            curr = curr.right;
        } else if (p.val < curr.val && q.val < curr.val) {
            curr = curr.left;
        } else {
            // Divergent path found, or we've hit p or q itself
            return curr;
        }
    }

    return null;
};

// Example Usage
//      6
//    /   \
//   2     8
//  / \   / \
// 0   4 7   9
//    / \
//   3   5
const n0 = new TreeNode(0);
const n2 = new TreeNode(2);
const n3 = new TreeNode(3);
const n4 = new TreeNode(4);
const n5 = new TreeNode(5);
const n6 = new TreeNode(6);
const n7 = new TreeNode(7);
const n8 = new TreeNode(8);
const n9 = new TreeNode(9);

n6.left = n2; n6.right = n8;
n2.left = n0; n2.right = n4;
n4.left = n3; n4.right = n5;
n8.left = n7; n8.right = n9;

console.log(lowestCommonAncestor(n6, n2, n8).val); // 6
console.log(lowestCommonAncestor(n6, n2, n4).val); // 2
